File: MetaGeta/MetaGeta.GUI/ViewModels/TvShowViewModel.vb
Imports System.Text.RegularExpressions
Imports System.Net.Cache
Imports System.Globalization
Imports System.Net

Public Class TvShowViewModel
    Inherits NavigationTab

    Private ReadOnly m_DataStore As MGDataStore
    Private m_Series As ObservableCollection(Of TvSeries)

    Public Sub New(ByVal dataStore As MGDataStore)
        m_DataStore = dataStore
        AddHandler m_DataStore.PropertyChanged, AddressOf DataStorePropertyChanged

        m_Series = CreateRuntimeData()
    End Sub

    Private Sub DataStorePropertyChanged(ByVal sender As Object, ByVal e As PropertyChangedEventArgs)
        If e.PropertyName = "ImportStatus" Then
            m_Series = CreateRuntimeData()
            OnPropertyChanged("Series")
        End If
    End Sub

    Public Overrides ReadOnly Property Caption() As String
        Get
            Return "TV Shows"
        End Get
    End Property

    Private Function CreateRuntimeData() As ObservableCollection(Of TvSeries)
        Dim results As New ObservableCollection(Of TvSeries)
        Dim seriesList = From f In m_DataStore.Files Group By SeriesName = f.GetTag(TVShowDataStoreTemplate.SeriesTitle) Into Group Order By SeriesName
        For Each seriesGroup In seriesList
            Dim series As New TvSeries(seriesGroup.SeriesName)
            series.SeriesBannerPath = seriesGroup.Group.Select(Function(file) file.GetTag(TVShowDataStoreTemplate.SeriesBanner)).Coalesce()
            series.SeriesPosterPath = seriesGroup.Group.Select(Function(file) file.GetTag(TVShowDataStoreTemplate.SeriesPoster)).Coalesce()

            Dim episodeGroups = From f In seriesGroup.Group Group By _
                                EpisodeNumber = f.GetTag(TVShowDataStoreTemplate.EpisodeNumber), _
                                SeasonNumber = f.GetTag(TVShowDataStoreTemplate.SeasonNumber) _
                                Into Group _
                                Select SeasonNumber, _
                                    EpisodeNumber, _
                                    Group, _
                                    FirstAired = Group.Select(Function(file) file.GetTag(TVShowDataStoreTemplate.EpisodeFirstAired)).Coalesce(), _
                                    Length = Group.Select(Function(file) file.GetTag(TVShowDataStoreTemplate.PlayTime)).Coalesce(), _
                                    Title = Group.Select(Function(file) file.GetTag(TVShowDataStoreTemplate.EpisodeTitle)).Coalesce()
            Dim episodes = From ep In episodeGroups _
                           Select New TvEpisode(series, ep.SeasonNumber, ep.EpisodeNumber, ep.Title) With { _
                               .AirDateString = ep.FirstAired, _
                               .LengthString = ep.Length _
                           }
            episodes = From ep In episodes Order By ep.SeasonNumber, ep.EpisodeNumber Select ep

            series.Episodes.AddRange(episodes)
            results.Add(series)
        Next
        Return results
    End Function


    Public ReadOnly Property Series() As ObservableCollection(Of TvSeries)
        Get
            Return m_Series
        End Get
    End Property

    Public Overrides ReadOnly Property Icon() As System.Windows.Media.ImageSource
        Get
            Return s_ViewImage
        End Get
    End Property

    Private Shared ReadOnly s_ViewImage As New BitmapImage(New Uri("pack://application:,,,/MetaGeta.GUI;component/Resources/view_detailed.png"))
End Class

Public Class DesignTimeTvShowViewModel
    Public ReadOnly Property Series() As ObservableCollection(Of TvSeries)
        Get
            Dim results As New ObservableCollection(Of TvSeries)
            results.Add(New TvSeries("Mythbusters") With {.SeriesBannerPath = "f:\ipod\73388-g.jpg", .SeriesPosterPath = "f:\ipod\img_posters_73388-1.jpg"})
            results.Last().Episodes.Add(New TvEpisode(results.Last(), 8, 1, "Demolition Derby"))
            results.Last().Episodes.Add(New TvEpisode(results.Last(), 8, 2, "Alaska Special 2"))
            results.Last().Episodes.Add(New TvEpisode(results.Last(), 8, 3, "Banana Slip/Double-Dip"))
            results.Add(New TvSeries("Arrested Development") With {.SeriesBannerPath = "f:\ipod\72173-g.jpg", .SeriesPosterPath = "f:\ipod\72173-1.jpg"})
            results.Last().Episodes.Add(New TvEpisode(results.Last(), 1, 1, "Pilot") With {.AirDate = New DateTime(2003, 11, 2), .Length = New TimeSpan(0, 22, 0)})
            results.Last().Episodes.Add(New TvEpisode(results.Last(), 1, 2, "Top Banana") With {.AirDate = New DateTime(2003, 11, 9), .Length = New TimeSpan(0, 22, 0)})
            results.Last().Episodes.Add(New TvEpisode(results.Last(), 1, 3, "Bringing Up Buster") With {.AirDate = New DateTime(2003, 11, 16), .Length = New TimeSpan(0, 22, 0)})
            Return results
        End Get
    End Property
End Class

Public Class TvSeries
    Private ReadOnly m_Name As String
    Private m_Episodes As New List(Of TvEpisode)
    Private m_SeriesBannerPath, m_SeriesPosterPath As String

    Public Sub New(ByVal seriesName As String)
        m_Name = seriesName
    End Sub

    Public ReadOnly Property Name() As String
        Get
            Return m_Name
        End Get
    End Property

    Public ReadOnly Property Episodes() As IList(Of TvEpisode)
        Get
            Return m_Episodes
        End Get
    End Property

    Public Property SeriesBannerPath() As String
        Get
            Return m_SeriesBannerPath
        End Get
        Set(ByVal value As String)
            m_SeriesBannerPath = value
        End Set
    End Property

    Public Property SeriesPosterPath() As String
        Get
            Return m_SeriesPosterPath
        End Get
        Set(ByVal value As String)
            m_SeriesPosterPath = value
        End Set
    End Property

    Public ReadOnly Property HasPoster() As Boolean
        Get
            Return SeriesBannerPath IsNot Nothing
        End Get
    End Property
End Class

Public Class TvEpisode
    Private ReadOnly m_Title As String
    Private m_AirDate As DateTime
    Private m_Length As TimeSpan
    Private ReadOnly m_Series As TvSeries
    Private ReadOnly m_SeasonNumber As Integer, m_EpisodeNumber As Integer


    Public Sub New(ByVal series As TvSeries, ByVal seasonNumber As String, ByVal episodeNumber As String, ByVal title As String)
        Me.New(series, 1, 1, title)
        If Not Integer.TryParse(seasonNumber, m_SeasonNumber) Then m_SeasonNumber = 0
        If Not Integer.TryParse(episodeNumber, m_EpisodeNumber) Then m_EpisodeNumber = 0
    End Sub

    Public Sub New(ByVal series As TvSeries, ByVal seasonNumber As Integer, ByVal episodeNumber As Integer, ByVal title As String)
        m_Series = series
        m_SeasonNumber = seasonNumber
        m_EpisodeNumber = episodeNumber
        m_Title = title
    End Sub

    Public ReadOnly Property Series() As TvSeries
        Get
            Return m_Series
        End Get
    End Property

    Public ReadOnly Property SeasonNumber() As Integer
        Get
            Return m_SeasonNumber
        End Get
    End Property

    Public ReadOnly Property EpisodeNumber() As Integer
        Get
            Return m_EpisodeNumber
        End Get
    End Property

    Public ReadOnly Property Title() As String
        Get
            Return m_Title
        End Get
    End Property

    Public Property AirDate() As Date
        Get
            Return m_AirDate
        End Get
        Set(ByVal value As Date)
            m_AirDate = value
        End Set
    End Property

    Public Property AirDateString() As String
        Get
            Return m_AirDate.ToString()
        End Get
        Set(ByVal value As String)
            If value IsNot Nothing Then
                m_AirDate = DateTime.ParseExact(value, "u", CultureInfo.CurrentCulture)
            End If
        End Set
    End Property

    Public Property Length() As TimeSpan
        Get
            Return m_Length
        End Get
        Set(ByVal value As TimeSpan)
            m_Length = value
        End Set
    End Property

    Public Property LengthString() As String
        Get
            Return m_Length.ToString()
        End Get
        Set(ByVal value As String)
            If value IsNot Nothing Then
                m_Length = TimeSpan.Parse(value)
            End If
        End Set
    End Property

    Public ReadOnly Property LongName() As String
        Get
            If SeasonNumber = 0 Then
                Return String.Format("{1:00} - {2}", EpisodeNumber, Title)
            Else
                Return String.Format("{0:00}x{1:00} - {2}", SeasonNumber, EpisodeNumber, Title)
            End If
        End Get
    End Property

    Public ReadOnly Property IsEven() As Boolean
        Get
            Return EpisodeNumber Mod 2 = 0
        End Get
    End Property
End Class

<ValueConversion(GetType(String), GetType(ImageSource))> _
Public Class UriImageConverter
    Implements IValueConverter

    Public Function Convert(ByVal value As Object, ByVal targetType As System.Type, ByVal parameter As Object, ByVal culture As System.Globalization.CultureInfo) As Object Implements System.Windows.Data.IValueConverter.Convert
        If value Is Nothing Then
            Return Nothing
        Else
            Try
                Return New BitmapImage(New Uri(CType(value, String)), New RequestCachePolicy(RequestCacheLevel.Default))
            Catch wex As WebException
                Return Nothing
            End Try
        End If
    End Function

    Public Function ConvertBack(ByVal value As Object, ByVal targetType As System.Type, ByVal parameter As Object, ByVal culture As System.Globalization.CultureInfo) As Object Implements System.Windows.Data.IValueConverter.ConvertBack
        Throw New NotImplementedException()
    End Function
End Class

File: MetaGeta/MetaGeta.GUI/ViewModels/TvShowViewModel.cs
﻿// Copyright 2009 Will Thomas
// 
// This file is part of MetaGeta.
// 
// MetaGeta is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// MetaGeta is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with MetaGeta. If not, see <http://www.gnu.org/licenses/>.

#region

using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.Globalization;
using System.Linq;
using System.Net;
using System.Net.Cache;
using System.Windows.Data;
using System.Windows.Media;
using System.Windows.Media.Imaging;
using MetaGeta.DataStore;
using MetaGeta.Utilities;
using GalaSoft.MvvmLight.Messaging;

#endregion

namespace MetaGeta.GUI {
	public class TvShowViewModel : NavigationTab {
		private static readonly BitmapImage s_ViewImage = new BitmapImage(new Uri("pack://application:,,,/MetaGeta.GUI;component/Resources/tv.png"));
		private readonly MGDataStore m_DataStore;

		private ObservableCollection<TvSeries> m_Series;

		public TvShowViewModel(NavigationTabGroupBase group, IMessenger messenger, MGDataStore dataStore)
			: base(group, messenger) {
			m_DataStore = dataStore;
			m_DataStore.PropertyChanged += DataStorePropertyChanged;

			m_Series = CreateRuntimeData();
		}

		public override string Caption { get { return "TV Shows"; } }
		public ObservableCollection<TvSeries> Series { get { return m_Series; } }
		public override ImageSource Icon { get { return s_ViewImage; } }

		private void DataStorePropertyChanged(object sender, PropertyChangedEventArgs e) {
			if (e.PropertyName == "ImportStatus") {
				m_Series = CreateRuntimeData();
				RaisePropertyChanged("Series");
			}
		}

		private ObservableCollection<TvSeries> CreateRuntimeData() {
			var results = new ObservableCollection<TvSeries>();
			IOrderedEnumerable<IGrouping<string, MGFile>> seriesList = from f in m_DataStore.Files
																	   group f by f.Tags.GetString(TVShowDataStoreTemplate.SeriesTitle)
																		   into asdf
																		   orderby asdf.Key
																		   select asdf;

			foreach (IGrouping<string, MGFile> seriesGroup in seriesList) {
				var series = new TvSeries(seriesGroup.Key);
				series.SeriesBannerPath = seriesGroup.Select(f => f.Tags.GetString(TVShowDataStoreTemplate.SeriesBanner)).Coalesce();
				series.SeriesPosterPath = seriesGroup.Select(f => f.Tags.GetString(TVShowDataStoreTemplate.SeriesPoster)).Coalesce();
				series.SeriesDescription = seriesGroup.Select(f => f.Tags.GetString(TVShowDataStoreTemplate.SeriesDescription)).Coalesce();

				var episodeGroups = from f in seriesGroup
									group f by
										new {
											EpisodeNumber = f.Tags.GetInt(TVShowDataStoreTemplate.EpisodeNumber),
											SeasonNumber = f.Tags.GetInt(TVShowDataStoreTemplate.SeasonNumber)
										}
										into grp
										select new {
											grp.Key.SeasonNumber,
											grp.Key.EpisodeNumber,
											Episodes = grp,
											FirstAired = grp.Select(file => file.Tags.GetDateTime(TVShowDataStoreTemplate.EpisodeFirstAired)).Coalesce(),
											EpisodeBannerPath = grp.Select(file => file.Tags.GetString(TVShowDataStoreTemplate.EpisodeBanner)).Coalesce(),
											Length = grp.Select(file => file.Tags.GetTimeSpan(TVShowDataStoreTemplate.PlayTime)).Coalesce(),
											Title = grp.Select(file => file.Tags.GetString(TVShowDataStoreTemplate.EpisodeTitle)).Coalesce(),
											Ipod = grp.Any(file => true == file.Tags.GetBool(TVShowDataStoreTemplate.iPodClassicCompatible)),
											Iphone = grp.Any(file => true == file.Tags.GetBool(TVShowDataStoreTemplate.iPhoneCompatible)),
											Computer = grp.Any(),
										};

				IEnumerable<TvEpisode> episodes = from ep in episodeGroups
												  where ep.SeasonNumber.HasValue && ep.EpisodeNumber.HasValue
												  select new TvEpisode(series, ep.SeasonNumber.Value, ep.EpisodeNumber.Value, ep.Title) {
													  AirDate = ep.FirstAired,
													  EpisodeBannerPath = ep.EpisodeBannerPath,
													  Length = ep.Length,
													  HasComputerVersion = ep.Computer,
													  HasIphoneVersion = ep.Iphone,
													  HasIpodVersion = ep.Ipod,
												  };
				episodes = from ep in episodes orderby ep.SeasonNumber, ep.EpisodeNumber select ep;

				series.Episodes.AddRange(episodes);
				results.Add(series);
			}
			return results;
		}
	}

	public class DesignTimeTvShowViewModel {
		public ObservableCollection<TvSeries> Series {
			get {
				var results = new ObservableCollection<TvSeries>();
				results.Add(new TvSeries("Mythbusters") {
					SeriesBannerPath = "e:\\ipod.not\\73388-g.jpg",
					SeriesPosterPath = "e:\\ipod.not\\img_posters_73388-1.jpg",
					SeriesDescription = "So exactly how hard is it to find a needle in a haystack, anyway? And can water dripping on your forehead really drive you nuts? Those are the kinds of questions, myths and urban legends that are put to the test in this humorous series to find out which ones hold water (and keep it from dripping on your forehead and driving you insane).",
				});
				results.Last().Episodes.Add(new TvEpisode(results.Last(), 6, 1, "James Bond Special Part 1") { EpisodeBannerPath = @"e:\ipod.not\Images\408062.jpg", });
				results.Last().Episodes.Add(new TvEpisode(results.Last(), 6, 2, "Lead Balloon") { EpisodeBannerPath = @"e:\ipod.not\Images\408062.jpg", });
				results.Last().Episodes.Add(new TvEpisode(results.Last(), 7, 1, "Demolition Derby") { EpisodeBannerPath = @"e:\ipod.not\Images\408062.jpg", });
				results.Last().Episodes.Add(new TvEpisode(results.Last(), 7, 2, "Alaska Special 2") { EpisodeBannerPath = @"e:\ipod.not\Images\408062.jpg", });
				results.Last().Episodes.Add(new TvEpisode(results.Last(), 7, 13, "Banana Slip/Double-Dip") { EpisodeBannerPath = @"e:\ipod.not\Images\408062.jpg", });
				results.Add(new TvSeries("Arrested Development") {
					SeriesBannerPath = "f:\\ipod\\72173-g.jpg",
					SeriesPosterPath = "f:\\ipod\\72173-1.jpg"
				});
				results.Last().Episodes.Add(new TvEpisode(results.Last(), 1, 1, "Pilot") {
					AirDate = new DateTime(2003, 11, 2),
					Length = new TimeSpan(0, 22, 0),
					EpisodeBannerPath = @"e:\ipod.not\Images\408062.jpg",
				});
				results.Last().Episodes.Add(new TvEpisode(results.Last(), 1, 2, "Top Banana") {
					AirDate = new DateTime(2003, 11, 9),
					Length = new TimeSpan(0, 22, 0),
					EpisodeBannerPath = @"e:\ipod.not\Images\408062.jpg",
				});
				results.Last().Episodes.Add(new TvEpisode(results.Last(), 1, 3, "Bringing Up Buster") {
					AirDate = new DateTime(2003, 11, 16),
					Length = new TimeSpan(0, 22, 0),
					EpisodeBannerPath = @"e:\ipod.not\Images\408062.jpg",
				});
				return results;
			}
		}
	}

	public class TvSeries {
		private readonly string m_Name;
		private readonly List<TvEpisode> m_Episodes = new List<TvEpisode>();

		public TvSeries(string seriesName) {
			m_Name = seriesName;
		}

		public string Name { get { return m_Name; } }
		public IList<TvEpisode> Episodes { get { return m_Episodes; } }

		public string SeriesBannerPath { get; set; }
		public string SeriesPosterPath { get; set; }
		public bool HasPoster { get { return SeriesBannerPath != null; } }

		public string SeriesDescription { get; set; }

		public bool HasMultipleSeasons { get { return Episodes.Select(e => e.SeasonNumber).Distinct().Count() > 1; } }
	}

	public class TvEpisode {
		private readonly int m_EpisodeNumber;
		private readonly int m_SeasonNumber;
		private readonly TvSeries m_Series;
		private readonly string m_Title;

		public TvEpisode(TvSeries series, int seasonNumber, int episodeNumber, string title) {
			m_Series = series;
			m_SeasonNumber = seasonNumber;
			m_EpisodeNumber = episodeNumber;
			m_Title = title;
		}

		public TvSeries Series { get { return m_Series; } }
		public int SeasonNumber { get { return m_SeasonNumber; } }
		public int EpisodeNumber { get { return m_EpisodeNumber; } }
		public string Title { get { return m_Title; } }
		public string EpisodeBannerPath { get; set; }
		public DateTimeOffset AirDate { get; set; }
		public TimeSpan Length { get; set; }

		public string LongName {
			get {
				if (SeasonNumber == 0)
					return string.Format("{1:00} - {2}", EpisodeNumber, Title);
				else
					return string.Format("{0:00}x{1:00} - {2}", SeasonNumber, EpisodeNumber, Title);
			}
		}

		public bool HasComputerVersion { get; set; }
		public bool HasIpodVersion { get; set; }
		public bool HasIphoneVersion { get; set; }

		public IEnumerable<Tuple<string, string>> Details {
			get {
				yield return new Tuple<string, string>("First Aired:", AirDate.ToString("d"));
				yield return new Tuple<string, string>("Length:", Length.ToString("g"));
			}
		}
	}

	[ValueConversion(typeof(string), typeof(ImageSource))]
	public class UriImageConverter : IValueConverter {
		#region IValueConverter Members

		public object Convert(object value, Type targetType, object parameter, CultureInfo culture) {
			if (value == null)
				return null;
			else {
				try {
					return new BitmapImage(new Uri((string)value), new RequestCachePolicy(RequestCacheLevel.Default));
				} catch (WebException wex) {
					return null;
				}
			}
		}

		public object ConvertBack(object value, Type targetType, object parameter, CultureInfo culture) {
			throw new NotImplementedException();
		}

		#endregion
	}
}